### ISSUE TITLE
[Bug] Trakt/Kodi not scrobbling/tracking local videos if played from TMDB Helper addon/widget

### ISSUE THREAD

**Issue by ShibajiCh** [jurialmunkey/plugin.video.themoviedb.helper#349]
Hi jurialmunkey,
I'm currently using version 4.0.36 of TMDB Helper with Arctic Zephyr 2 skin on Kodi version 18.9. I noticed that if I play any local video from the video info screen then the video plays fine but Trakt doesn't scrobble/track that particular video and even Kodi doesn't track its progress. It appears as if Kodi/Trakt can't detect the video being played at all. However, if I click on the context menu and then click on "Select Player" to play the video via the default Kodi player then Trakt scrobbles/tracks the video perfectly and Kodi also tracks its progress. Also, Trakt scrobbling/tracking works fine for non-local videos player through Seren (Source Select) player. I checked on multiple devices. I guess there's some issue in the settings for the Kodi (default player) in TMDB Helper. Earlier you changed the is_resolvable field for the Kodi (default player) in TMDB Helper and fixed another similar issue (please refer issue #332). May be this is also something similar to that. Could you please check what's the cause of this issue? Thanks!

**Comment by jurialmunkey** [comment]
Are these "true" local files (e.g. exists on hdd/network and just scanned to Kodi library) or do they first go via some other addon (e.g. exist in Plex and added to local library via PKC)?

Also, where are you opening the info dialog from? Do mean that you are opening info on an item in the Kodi library itself or are you opening info on an item from a TMDbHelper folder/widget that also happens to be in the local library?

**Comment by ShibajiCh** [comment]
Hi jurialmunkey,
Yes, these are "true" local files, i.e. they exist on HDD/network and are scanned to Kodi library.

I'd like to add that I checked again and found that this issue is not only happening from within the video info dialog but also from the "Play" button within the lists inside the TMDB Helper addon. It's happening whether I use TMDB Helper folder/widget, but only for local library video files. Non-local video files if played from Seren (Source Select) player have no issues. Also, if I use the "Select Player" option on the TMDB Helper context menu and play the same local library video file on the default Kodi player then also there is no issue. Kodi library video files played without using the TMDB Helper addon also have no issues.

What I can see is that when I play any local library video file using the TMDB Helper it's progress is tracked only by the TMDB Helper addon. Neither Trakt nor Kodi tracks it's progress, i.e. it's play status in not tracked by Trakt and also Kodi doesn't add it to it's default "In Progress" list. So it's definitely a TMDB Helper addon issue as I checked on different Kodi setups on multiple devices and found the same issue on each one of them. At present, the only solution to this issue as I mentioned above is to play the local library video on the default Kodi player using the "Select Player" option on the TMDB Helper context menu.

Please advise. Thanks!

**Comment by ShibajiCh** [comment]
Hi jurialmunkey,
Any updates on this issue? Please advise. Thanks!

**Comment by jurialmunkey** [comment]
This is working fine for me. I just tested and I can't recreate this issue.

Do your local items scrobble correctly if you just play them directly from the library?

Also, maybe a stupid question, but do you have the Trakt addon installed? Just thought it worthwhile to point out that TMDbHelper doesn't scrobble *any* content itself. You need to have the Trakt addon installed to scrobble local content (and for other players it is up to the individual plugin being called to scrobble content).

**Comment by ShibajiCh** [comment]
Hi jurialmunkey,
Yes, my local items scrobble correctly if I just play them directly from the library. I also have the Trakt addon installed. Please note that the issue is not only with Trakt scrobbling but the progress is not being tracked by Kodi as well, i.e.the local video being played is not shown under Kodi in progress items as well. However, if I use "Select Player" option from TMDB Helper context menu and play the local video file in the Kodi default player then everything is working fine. Earlier in TMDB Helper version 3.x this issue was not there. Also I replicated this issue in 3 different Kodi setups on 3 different devices.

**Comment by jurialmunkey** [comment]
I honestly can't recreate this issue. I tried in both Arctic Horizon and Estuary and everything I played scrobbled correctly and updated progress in Kodi library. I played local episodes both directly and from info dialog using a variety of home widgets and also inside TMDbHelper -- I was really trying to break it and couldn't.

Have you tried in Estuary to rule out a skin issue?

**Comment by ShibajiCh** [comment]
Hi jurialmunkey,
That's weird! I tried playing local video files from within the TMDB Helper addon in Estuary Skin as well in my Windows PC and the same issue happens. I also tried in Arctic Zephyr 2 Skin on my 2 Fire Stick 4Ks and the same issue happens. Can't understand why?

**Comment by jurialmunkey** [comment]
> Is there any Kodi default player configuration that I need to change?

Nope, there's no user configuration of local player. The option for the local player is enabling auto play of local files (basically like setting local to be the default player if available) - but that shouldn't matter (I tested with it on/off just to make sure and it worked fine).

> Note: I'm using Tmdb Helper version 4.0.36. Also I tried playing local movies and not episodes.

Can you test with 4.0.41 (latest git master) just to make sure I haven't inadvertently fixed this issue with some other change that I made in the last couple of days.

**Comment by jurialmunkey** [comment]
>  Also I tried playing local movies and not episodes.

Ah wait, yeah I think you're right. I was testing episodes which were working fine but I just tried a movie and it isn't being picked up.

**Comment by ShibajiCh** [comment]
Hi jurialmunkey,
Yes, that's what I was telling you about. I shall wait for your further update on this issue.

**Comment by jurialmunkey** [comment]
Okay I've got Kodi's local resume/progress working properly now for movies but still struggling to get them recognised by the Trakt addon. I can't figure out what field it uses to trigger the scrobble. I'm going to have a dig around in its code to see if I can find what it is looking for.

**Comment by ShibajiCh** [comment]
Hi jurialmunkey,
Thank you! I shall wait for your further update regarding the Trakt issue.

**Comment by ShibajiCh** [comment]
Hi jurialmunkey,
I updated TMDB Helper to version 4.0.42 using the latest Git Master file. However, for me Kodi's local resume/progress is not working properly. Kodi is not adding the video to In Progress list even after playing half of the video. Please advise.

**Comment by jurialmunkey** [comment]
> However, for me Kodi's local resume/progress is not working properly. Kodi is not adding the movie to the In Progress list even after playing half of the video. 

No, I just meant resume from inside TMDbHelper (i.e. player local file via TMDbHelper will save resume point so next time you play from TMDbHelper will ask to resume from there).

Latest version now has option in TMDbHelper's settings under Accounts for a hack to enable Trakt scrobbling - this will set resume points in the library but will break the resume points in TMDbHelper. I can't do better than this - it is related to a bug in Kodi core where mediatype is not being set correctly and also the way the Trakt addon manages things (same reason it doesn't work with addons).

**Comment by ShibajiCh** [comment]
Hi jurialmunkey,
Thanks for fixing the Trakt Scrobbling issue!
It's working fine. However, like you mentioned above, the hack has broken the resume points in TMDB Helper for local files, but I can live with it as I can find the files under In Progress section in Kodi.

Now coming to another issue. I noticed that enabling the hack also broke the resume points for non-local files in TMDB Helper as well. For example, if I'm playing a non-local file in TMDB Helper using the Play button or "Select Player" option and choosing "Seren (Source Select)" player then the resume point is no longer tracked by the TMDB Helper addon. Is there any solution to this for non-local files only or is the hack going to break both local and non-local file resume points in TMDB Helper? Please advise.

NOTE: I can find the non-local files within "Finish Watching" list under "Movies" in the Seren addon though, as it is being scrobbled by Trakt after using the hack.

**Comment by jurialmunkey** [comment]
Kodi will only set resume points to listitems if a playable URL is passed back to Kodi via the setResolvedURL method. Since you are using PlayMedia for that particular player, Kodi won't set resume/watched status.

In a really simplified sense, setResolvedURL basically attaches the final playable url back to the original listitem whereas PlayMedia plays the url in separate unrelated process (so no item to attach the url to and thus nothing to attach progress to). The reason some plugins don't work properly with setResolvedURL is because the URL they provide doesn't resolve to a playable file and is instead a callback function to some other scripted process in the plugin (e.g. scraping sources).

Also, non-local files aren't scrobbled by the Trakt addon. They are scrobbled by whatever addon plays them.

**Comment by ShibajiCh** [comment]
Hi jurialmunkey,
Okay, thanks for explaining in details.
Just one quick question, if Seren was scrobbling the non-local files, then how was TMDB Helper able to resume them when the Trakt hack was not used?

**Comment by jurialmunkey** [comment]
> Just two quick questions, if Seren was scrobbling the non-local files, then how was TMDB Helper able to resume them when the Trakt hack was not used?

Seren probably caches a resume point independently from Kodi. PlayMedia is practically the same as clicking on the item in the actual plugin itself - Simplified, TMDbHelper just walks through the plugin's directories virtually according to the rules defined in the player. Once it finds a match, it gives the item's URL to PlayMedia, the plugin then runs and does whatever code is associated with the parameters in the URL. Point is that it isn't TMDbHelper or Kodi that is resuming, but the end plugin that is handling that in this case.

Also note the Trakt hack is *only* applied to local files (i.e. URLs that don't start with a plugin:// callback). It isn't ever applied to non-local files - the code remains exactly the same for those.

> And now that the hack is enabled why can't TMDB Helper no longer create the resume points for these non-local files even though Seren is scrobbling them?

Because Seren is only saving resume points for items that are played via Seren. The resume point was never created for the TMDbHelper item if you played it via PlayMedia. The resume point was only attached to the TMDbHelper item if you played the local file (because it was played via setResolvedURL which attaches its progress back to the listitem itself).

**Comment by ShibajiCh** [comment]
Hi, jurialmunkey,
That's weird! Actually I thought that I noticed a few days back when I played a movie via the "Seren (Source Select)" player in TMDB Helper, then stopped it in the middle and again played it back, it resumed from the position where I stopped earlier. However, may be this is not the way it happened and I might be thinking about it all wrong. Nevertheless, thanks for the Trakt hack! It has resolved majority of the problems that I was facing.
